featurebyte/service/preview.py

```diff
@@ -152,6 +167,55 @@ async def _get_or_cache_dataframe(
             session=session, feature_store_id=feature_store_id, query=query
         )
 
+    async def _get_graph_using_sampled_primary_table(
+        self,
+        feature_store: FeatureStoreModel,
+        session: BaseSession,
+        sample: FeatureStoreSample,
+        num_rows: int,
+        seed: int,
+        total_num_rows: Optional[int],
+    ) -> GraphWithSampledPrimaryTable:
+        # get primary table node & construct a new graph for sampling
+        primary_table_input_node = sample.graph.get_sample_table_node(node_name=sample.node_name)
+        graph = QueryGraph()
+        inserted_input_node = graph.add_node(primary_table_input_node, input_nodes=[])
+        graph_interpreter = GraphInterpreter(
+            query_graph=graph, source_info=feature_store.get_source_info()
+        )
+        graph_info = graph_interpreter.extract_graph_info_for_sampling(
+            query_graph=sample.graph,
+            target_node_name=sample.node_name,
+        )
+        num_rows = int(num_rows * graph_info.get_oversampling_factor())
+        sample_sql_tree, _ = graph_interpreter._construct_sample_sql(
+            node_name=inserted_input_node.name,
+            num_rows=num_rows,
+            seed=seed,
+            from_timestamp=sample.from_timestamp,
+            to_timestamp=sample.to_timestamp,
+            timestamp_column=sample.timestamp_column,
+            total_num_rows=total_num_rows,
+            skip_conversion=True,  # skip conversion to keep original table column types
+            sample_on_primary_table=False,  # go through normal sampling process with single table
+        )
+        sampled_table_name, is_table_cached = await self._get_or_cache_table(
+            session=session,
+            params=sample,
+            table_expr=sample_sql_tree,
+        )
+        output = sample.graph
+        for node in output.nodes:
+            if node.name == primary_table_input_node.name:
+                assert isinstance(node, InputNode)
+                node.parameters.table_details = TableDetails(table_name=sampled_table_name)
```

Comment:
This is mutating the graph passed in, right? Just checking this is the intended behaviour, as the resulting graph will not be valid for other purposes other than being used by this service. I'm mainly concerned about the usage in feature ideation.